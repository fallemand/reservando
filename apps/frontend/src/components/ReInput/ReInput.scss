@import "~@/assets/styles/variables";

$padding-y: 0.5em;
$padding-x: 0.75em;

.re-input {
  display: flex;
  position: relative;
  color: $grape;

  &__field {
    padding: $padding-y $padding-x;
    width: 100%;
    line-height: normal;
    border: 0;
    border-bottom: 2px solid $grape;
    font-weight: 600;
    text-align: center;
    font-size: 2em;
    font-family: inherit;
    color: inherit;
    background-color: transparent;
    outline: 0;
    letter-spacing: -1px;

    &::placeholder {
      color: $input-placeholder;
    }
  }

  &__icon {
    position: absolute;
    display: flex;
    align-items: center;
    height: 100%;
    color: inherit;
    pointer-events: none;

    &--preicon {
      left: $padding-x;

      & ~ .re-input__field {
        padding-left: $padding-x * 3;
      }
    }

    &--posticon {
      right: $padding-x;

      & ~ .re-input__field {
        padding-right: $padding-x * 3;
      }
    }
  }
}
