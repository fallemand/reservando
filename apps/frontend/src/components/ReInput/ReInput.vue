<template>
  <div class="re-input">
    <span v-if="$slots.preicon" class="re-input__icon re-input__icon--preicon">
      <slot name="preicon" />
    </span>
    <span v-if="$slots.posticon" class="re-input__icon re-input__icon--posticon">
      <slot name="posticon" />
    </span>
    <input
      ref="input"
      class="re-input__field"
      v-bind="$attrs"
      v-on="{
        change: emitChange,
        input: emitInput,
      }"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";

const ReInput = defineComponent({
  name: "ReInput",
  inheritAttrs: false,
  methods: {
    emitChange($evt: Event): void {
      /** `(value: string, $evt: Event)` */
      this.$emit("change", ($evt.target as HTMLInputElement).value, $evt);
    },
    emitInput($evt: Event): void {
      /** `(value: string, $evt: Event)` */
      this.$emit("input", ($evt.target as HTMLInputElement).value, $evt);
    },
  },
});

export default ReInput;
</script>

<style lang="scss" src="./ReInput.scss" />
