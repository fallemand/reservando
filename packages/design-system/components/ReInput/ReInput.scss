@import "../../styles/variables";

$padding-y: 0.5em;
$padding-x: 0.75em;

.re-input {
  display: flex;
  position: relative;
  color: $grape;
  font-weight: 500;

  &__field {
    padding: $padding-y $padding-x;
    width: 100%;
    line-height: normal;
    font-size: 1em;
    font-weight: inherit;
    text-align: inherit;
    font-family: inherit;
    color: inherit;
    outline: 0;

    &--underline {
      border: 0;
      border-bottom: 2px solid $grape;
      background-color: transparent;
    }

    &--outline {
      border: 1px solid $border-primary;
      border-radius: $bdu * 1.5;
      background-color: $white;
    }

    &::placeholder {
      color: $text-secondary;
    }
  }

  &__icon {
    position: absolute;
    display: flex;
    align-items: center;
    height: 100%;
    color: inherit;
    pointer-events: none;

    &--preicon {
      left: $padding-x;

      & ~ .re-input__field {
        padding-left: $padding-x * 3;
      }
    }

    &--posticon {
      right: $padding-x;

      & ~ .re-input__field {
        padding-right: $padding-x * 3;
      }
    }
  }
}
