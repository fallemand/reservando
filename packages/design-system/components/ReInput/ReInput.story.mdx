import { Meta, Story, Canvas, Props } from "@storybook/addon-docs/blocks";
import { ReInput, ReButton, ReIcon } from "@reservando/design-system";

<Meta title="Components/Input" component={ReInput} />

#### ReInput

<Canvas>
  <Story name="basic">
    {{
      components: { ReInput },
      data() {
        return {
          text: "",
        };
      },
      template: `
      <div style="max-width: 400px;">
        <ReInput v-model="text" placeholder="Placeholder" />
        <p class="hint mt-3">Value: {{ text }}</p>
      </div>
    `,
    }}
  </Story>
</Canvas>

## How to use

```html
<script lang="ts">
  import { ReInput } from "@reservando/design-system";
</script>
```

<ArgsTable of={ReInput} />

#### Input Attributes

Pass any input attribute `id`, `name`, `type`, `placeholder`, `value`, `required`, `minlength`, etc.

<Canvas>
  <Story name="input attributes">
    {{
      components: { ReInput },
      template: `
      <div style="max-width: 400px;">
        <ReInput class="mb-3" value="I have a value here" />
        <ReInput class="mb-3" type="password" placeholder="Password here" />
        <ReInput class="mb-3" type="number" placeholder="Only numbers from 0 to 10" :max="10" :min="0" />
        <ReInput class="mb-3" placeholder="Im disabled" :disabled="true" />
      </div>
    `,
    }}
  </Story>
</Canvas>

#### Using other events

Pass any other input event `@mouseenter`, `@mouseleave`, `@keypressed`, etc. the same way you will do it with a `HTML input`

<Canvas>
  <Story name="html events">
    {{
      components: { ReInput },
      methods: {
        handleFocus($evt) {
          $evt.target.value = "Thanks for doing that";
        },
        handleEnter($evt) {
          alert("Awesome");
        },
        handleMouseOver($evt) {
          $evt.target.value = "Don't do it any more!";
        },
        handleMouseLeave($evt) {
          $evt.target.value = "Again again!!";
        },
      },
      template: `
      <div style="max-width: 400px;">
        <ReInput class="mb-3" placeholder="Focus me" @focus="handleFocus" />
        <ReInput class="mb-3" placeholder="Press Enter here" @keydown.enter="handleEnter" />
        <ReInput class="mb-3" placeholder="Hover me" @mouseover="handleMouseOver" @mouseleave="handleMouseLeave"/>
      </div>
    `,
    }}
  </Story>
</Canvas>

## Examples

#### Passing an icon

<Canvas>
  <Story name="with icon">
    {{
      components: { ReInput, ReIcon },
      template: `
      <div style="max-width: 400px;">
        <ReInput class="mb-3" placeholder="With preicon">
          <template #preicon><ReIcon name="plus" /></template>
        </ReInput>
        <ReInput class="mb-3" placeholder="With posticon">
          <template #posticon><ReIcon name="plus" /></template>
        </ReInput>
      </div>
    `,
    }}
  </Story>
</Canvas>

#### Using ref to access the HTML Input

Use `$refs.{refName}.$refs.input` to access the `Input element`

<Canvas>
  <Story name="using ref">
    {{
      components: { ReInput, ReButton },
      methods: {
        checkValidity() {
          alert(`valid: ${this.$refs.address.$refs.input.checkValidity()}`);
        },
      },
      template: `
      <div style="display: flex;">
        <ReInput ref="address" required placeholder="Required input" />
        <ReButton class="ml-3" @click="checkValidity">Check Validity</ReButton>
      </div>
    `,
    }}
  </Story>
</Canvas>
