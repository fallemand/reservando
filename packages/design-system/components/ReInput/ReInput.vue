<template>
  <div class="re-input" :class="$attrs.class">
    <span v-if="$slots.preicon" class="re-input__icon re-input__icon--preicon">
      <slot name="preicon" />
    </span>
    <span v-if="$slots.posticon" class="re-input__icon re-input__icon--posticon">
      <slot name="posticon" />
    </span>
    <input
      ref="input"
      class="re-input__field"
      :class="`re-input__field--${modifier}`"
      v-bind="{
        ...$attrs,
        class: undefined,
      }"
      v-on="{
        change: emitChange,
        input: emitInput,
      }"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";
export const MODIFIERS = ["underline", "outline"];

const ReInput = defineComponent({
  name: "ReInput",
  inheritAttrs: false,
  props: {
    modifier: {
      default: "underline",
      type: String as PropType<typeof MODIFIERS[number]>,
      validator: (value: string): boolean => MODIFIERS.includes(value),
    },
  },
  methods: {
    emitChange($evt: Event): void {
      /** `(value: string, $evt: Event)` */
      this.$emit("change", ($evt.target as HTMLInputElement).value, $evt);
    },
    emitInput($evt: Event): void {
      /** `(value: string, $evt: Event)` */
      this.$emit("input", ($evt.target as HTMLInputElement).value, $evt);
    },
  },
});

export default ReInput;
</script>

<style lang="scss" src="./ReInput.scss" />
