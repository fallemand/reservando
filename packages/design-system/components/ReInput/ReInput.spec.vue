<script lang="ts">
import { shallowMount } from "@vue/test-utils";
import ReInput from "./ReInput.vue";

describe("ReInput component", () => {
  const props = {
    modelValue: "",
  };
  it("should render with default props", () => {
    const wrapper = shallowMount(ReInput, {
      props,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render with preicon", () => {
    const wrapper = shallowMount(ReInput, {
      props,
      slots: {
        preicon: "__PREICON__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render with posticon", () => {
    const wrapper = shallowMount(ReInput, {
      props,
      slots: {
        posticon: "__POSTICON__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should call @change with args (value)", () => {
    const attrs = {
      onChange: jest.fn(),
    };
    const wrapper = shallowMount(ReInput, {
      props: {
        modelValue: "__VALUE__",
      },
      attrs,
    });
    const input = wrapper.find("input");
    input.trigger("change");
    input.trigger("input");
    expect(attrs.onChange).toHaveBeenCalledWith("__VALUE__");
  });

  it("should bind all attributes to the input", () => {
    const wrapper = shallowMount(ReInput, {
      attrs: {
        type: "number",
        min: 0,
      },
      props,
    });
    const input = wrapper.find("input");
    expect(input.attributes().type).toEqual("number");
    expect(input.attributes().min).toEqual("0");
  });

  it("should bind all listeners to the input", () => {
    const attrs = {
      onClick: jest.fn(),
      onKeydown: jest.fn(),
      onBlur: jest.fn(),
    };
    const wrapper = shallowMount(ReInput, {
      attrs,
      props,
    });
    const input = wrapper.find("input");
    input.trigger("click");
    input.trigger("keydown");
    input.trigger("blur");
    expect(attrs.onClick).toHaveBeenCalled();
    expect(attrs.onKeydown).toHaveBeenCalled();
    expect(attrs.onBlur).toHaveBeenCalled();
  });
});
</script>
