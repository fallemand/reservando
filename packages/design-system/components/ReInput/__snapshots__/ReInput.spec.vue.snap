// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReInput component should render with default props 1`] = `
<div
  class="re-input"
>
  <!--v-if-->
  <!--v-if-->
  <input
    class="re-input__field re-input__field--outline"
  />
</div>
`;

exports[`ReInput component should render with posticon 1`] = `
<div
  class="re-input"
>
  <!--v-if-->
  <span
    class="re-input__icon re-input__icon--posticon"
  >
    
    __POSTICON__
    
  </span>
  <input
    class="re-input__field re-input__field--outline"
  />
</div>
`;

exports[`ReInput component should render with preicon 1`] = `
<div
  class="re-input"
>
  <span
    class="re-input__icon re-input__icon--preicon"
  >
    
    __PREICON__
    
  </span>
  <!--v-if-->
  <input
    class="re-input__field re-input__field--outline"
  />
</div>
`;
